server/app/views/components/LinkElement.java

```diff
@@ -106,79 +108,96 @@ public LinkElement opensInNewTab() {
     return this;
   }
 
-  public ContainerTag asAnchorText() {
-    ContainerTag tag = Strings.isNullOrEmpty(href) ? div(text) : a(text).withHref(href);
-    return tag.withCondId(!Strings.isNullOrEmpty(id), id)
-        .withCondHref(!Strings.isNullOrEmpty(href), href)
-        .withCondTarget(doesOpenInNewTab, "_blank")
-        .withClasses(DEFAULT_LINK_STYLES, styles);
+  public DivTag asAnchorText() {
+    DivTag tag;
+    ATag atag;
+    if (Strings.isNullOrEmpty(href)) {
+      tag = div(text);
+      tag.withClasses(DEFAULT_LINK_STYLES, styles);
+      if (doesOpenInNewTab) {
+        tag.attr("target", "_blank");
+      }
+      return tag;
+    } else {
+      atag = a(text).withHref(href);
+      atag.withId(id).withClasses(DEFAULT_LINK_STYLES, styles);
+      if (doesOpenInNewTab) {
+        atag.attr("target", "_blank");
+      }
+      atag.attr("href", href);
+
+      tag = div().with(atag);
+
+      return tag;
+    }
```

Comment:
Not sure I know how to deal with this, but I am sure I didn't do it correctly as of now. The problem is that it either returns a `ATag` or a `DivTag` depending on whether there is an `href` defined. As of now it wraps the `ATag` with a `DivTag` and that causes problems for the `withTarget(..)` method

Maybe I can refactor it to split into two kinds of methods, one returning an `ATag` if there is an `href`, and the other returning a `DivTag` if there isnt. The caller should be able to tell which to use

Reply:
This might be a case where the method could return a more general type (e.g. `ContainerTag`). Then there'd be no need to wrap in a div. This is a case where I'd prefer we not have to restructure the DOM in order to return a consistent type. Downstream callers of `asAnchorText` hopefully won't need to care about the details of what the tag type actually is.

Reply:
I'm kind of snagged on an error [here](https://github.com/seattle-uat/civiform/blob/122815a732b8f5962894e4efb0c45c8161815332/server/app/views/applicant/ApplicantProgramSummaryView.java#L210)

![error_3_trace](https://user-images.githubusercontent.com/7295000/172475876-c01f3a79-b4a9-4a6f-9693-e554f3b81369.png)

Reply:
I tried doing similarly with a minimal example and it worked:

![error_3_works_minexample](https://user-images.githubusercontent.com/7295000/172476364-69a78d2b-02b7-4ab0-ae2e-3641b8a58b35.png)

Reply:
Oh nevermind. I reproduced the problem in the minimal example

![error_3_minexample_reproduced_trace](https://user-images.githubusercontent.com/7295000/172477366-9448092a-6584-4d70-85e7-640bb6e3e913.png)


![error_3_minexample_reproduced](https://user-images.githubusercontent.com/7295000/172477337-14c15af9-bd2c-47d2-9719-9f4a4ac02f8a.png)

Reply:
@clouser2 what do you think of this approach? Instead of returning either an `a` or a `div` from the [asAnchorText](https://github.com/seattle-uat/civiform/blob/122815a732b8f5962894e4efb0c45c8161815332/server/app/views/components/LinkElement.java#L113-L119), or places like [here](https://github.com/seattle-uat/civiform/blob/122815a732b8f5962894e4efb0c45c8161815332/server/app/views/admin/programs/ProgramAdministratorProgramListView.java#L169-L185), we return an `Optional<ATag>` instead of either return an `ATag` or a `DivTag`?

I guess I'm not sure how an anchor tag can be valid if it doesn't have an `href`. As such, why potentially return a `div` from the `asAnchorText` method?

It might take a bit of work to try this out and see if the browser tests still pass

Reply:
Honestly, I think this should probably be future work. The browser tests had passed at some point, so maybe this should be a future PR

Reply:
It looks like you're running into a situation where our `LinkElement` abstraction definitely has some usability issues. `asAnchorText` previously returned either 1) an `<a>` tag if an HREF attribute was provided or 2) a `<div>` containing the link text, styled to look like a link (but not actually clickable).

This is definitely a strange behavior and not worth trying to change as part of this upgrade. Due to this, we don't want downstream callers to assume that it's always an `<a>` tag. I think this is a case for having the following method signature in `LinkElement`:

`ContainerTag<?> asAnchorText()`

Downstream callers could then use `ContainerTag<?>` as the return type. Then only methods available for all container elements would be available (namely `.with`). Any attempts to call `.setHref` / `.setTarget` would actually indicate a leakage of abstraction. Furthermore, those could call the corresponding `LinkElement.setHref` / `LinkElement.opensInNewTab` methods rather than calling them on the returned element.

I still think that we want to avoid restructuring the DOM due to the way that J2HTML works. In my opinion, any instances where this is occurring are likely code smells where the previous application code might not have actually been working as we intended (e.g. calling `.withHref` on a `<div>` tag would have had no practical effect.

Reply:
Ok, thats helpfull. Yeah adding the `<?>` to `ContainerTag` definitely fixes the compile error. Thank you